### PR TITLE
fix(tracing): correct matching ot baggage headers

### LINKED ISSUE #9156: Zipkin plugin cannot parse OT baggage headers
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

Kong 2.8.1

### Current Behavior

In the function `parse_baggage_headers()` in `kong/tracing/propagation.lua`, the call `match(k, header_pattern)` is used to check if a header name matches a fixed pattern. For OT baggage headers, the header pattern used is `"^ot-baggage%-(.*)$"`.

The problem here is that the first hyphen in this pattern is not escaped using `%`, and in Lua's pattern strings, an unescaped hyphen matches zero or more occurrences of the preceding character(s) (essentially like `*` in regular expressions). This means that `obaggage-foo`, `otbaggage-foo`, and `ottttttttbaggage-foo` all match, but `ot-baggage-foo` does not. This would mean that parsing of `ot-baggage-*` headers is currently broken.

I haven't had time to verify this in Kong itself, but the `luajit` runtime at least confirms this:
```
> print(string.match("ot-baggage-foo", "^ot-baggage%-(.*)$"))
nil
> print(string.match("obaggage-foo", "^ot-baggage%-(.*)$"))
foo
> print(string.match("ottttttttbaggage-foo", "^ot-baggage%-(.*)$"))
foo
> print(string.match("ot-baggage-foo", "^ot%-baggage%-(.*)$"))
foo
```

This can be easily solved by escaping the hyphen, although performance-wise it might be even better to do a literal search instead (e.g. `string.find("ot-baggage-foo", "ot-baggage-", 1, true)`, passing `true` for the `plain` parameter); this way we avoid having to recompile the same pattern for every header in every request.

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Anything else?

This probably deserves a separate issue, but the fact that there is zero documentation regarding the Zipkin plugin's ability to parse and output baggage headers is a huge oversight.

### PR DESCRIPTION
**Pull request by pathak1515**
### Summary

In the function parse_baggage_headers() in kong/tracing/propagation.lua, the call match(k, header_pattern) is used to check if a header name matches a fixed pattern. For OT baggage headers, the header pattern used is "^ot-baggage%-(.*)$".

To fix this changed OT_BAGGAGE_PATTERN
### Full changelog

Update OT_BAGGAGE_PATTERN from `^ot-baggage%-(.*)$` to `^ot%-baggage%-(.*)$`

### Issue reference

Fix #9156

### CODE REVIEW

**Comment by pathak1515** [comment]
@mayocream Can you please approve the workflows?

File: kong/tracing/propagation.lua [hunk 1]
```diff
@@ -41,7 +42,7 @@ local function parse_baggage_headers(headers, header_pattern)
   -- account for both ot and uber baggage headers
   local baggage
   for k, v in pairs(headers) do
-    local baggage_key = match(k, header_pattern)
+    local baggage_key = find(k, header_pattern,1, true)
```
Comment (chronolaw) [inline kong/tracing/propagation.lua:45]: style issue, should be `, 1, true`

File: kong/tracing/propagation.lua [hunk 2]
```diff
@@ -124,7 +125,7 @@ local function parse_zipkin_b3_headers(headers, b3_single_header)
 
   local trace_id_header = headers["x-b3-traceid"]
   if trace_id_header and ((#trace_id_header ~= 16 and #trace_id_header ~= 32)
-                           or trace_id_header:match("%X")) then
+    or trace_id_header:match("%X")) then
```
Comment (chronolaw) [inline kong/tracing/propagation.lua:128]: I think this line should algin to `((#trace_id_header ...`, just like the original form.

**Comment by pathak1515** [comment]
@mayocream Not able to relate CI failures with the changes

File: kong/tracing/propagation.lua (hunk 1 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: Resolved

File: kong/tracing/propagation.lua (hunk 2 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: resolved

File: kong/tracing/propagation.lua [hunk 3]
```diff
@@ -14,7 +15,7 @@ local baggage_mt = {
 }
 
 local B3_SINGLE_PATTERN =
-  "^(%x+)%-(%x%x%x%x%x%x%x%x%x%x%x%x%x%x%x%x)%-?([01d]?)%-?(%x*)$"
+"^(%x+)%-(%x%x%x%x%x%x%x%x%x%x%x%x%x%x%x%x)%-?([01d]?)%-?(%x*)$"
```
Comment (mayocream) [inline kong/tracing/propagation.lua:18]: The indent change seems not to be needed.

File: kong/tracing/propagation.lua [hunk 4]
```diff
@@ -97,7 +98,7 @@ local function parse_zipkin_b3_headers(headers, b3_single_header)
 
     else
       trace_id, span_id, sampled, parent_id =
-        match(b3_single_header, B3_SINGLE_PATTERN)
+      match(b3_single_header, B3_SINGLE_PATTERN)
```
Comment (mayocream) [inline kong/tracing/propagation.lua:101]: ditto

File: kong/tracing/propagation.lua [hunk 5]
```diff
@@ -422,9 +423,9 @@ local function set(conf_header_type, found_header_type, proxy_span, conf_default
   -- If conf_header_type is set to `preserve`, found_header_type is used over default_header_type;
   -- if conf_header_type is set to `ignore`, found_header_type is not set, thus default_header_type is used.
   if conf_header_type ~= "preserve" and
-     conf_header_type ~= "ignore" and
-     found_header_type ~= nil and
-     conf_header_type ~= found_header_type
+  conf_header_type ~= "ignore" and
+  found_header_type ~= nil and
+  conf_header_type ~= found_header_type
```
Comment (mayocream) [inline kong/tracing/propagation.lua:428]: ditto

File: kong/tracing/propagation.lua [hunk 6]
```diff
@@ -448,9 +449,9 @@ local function set(conf_header_type, found_header_type, proxy_span, conf_default
 
   if conf_header_type == "b3-single" or found_header_type == "b3-single" then
     set_header("b3", to_hex(proxy_span.trace_id) ..
-        "-" .. to_hex(proxy_span.span_id) ..
-        "-" .. (proxy_span.should_sample and "1" or "0") ..
-        (proxy_span.parent_id and "-" .. to_hex(proxy_span.parent_id) or ""))
+    "-" .. to_hex(proxy_span.span_id) ..
+    "-" .. (proxy_span.should_sample and "1" or "0") ..
+    (proxy_span.parent_id and "-" .. to_hex(proxy_span.parent_id) or ""))
```
Comment (mayocream) [inline kong/tracing/propagation.lua:454]: ditto

**Review by mayocream — CHANGES_REQUESTED** [review]

**Comment by mayocream** [comment]
@pathak1515 Could you please add a test for it? thanks!

File: kong/tracing/propagation.lua (hunk 3 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: fixed

File: kong/tracing/propagation.lua (hunk 4 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: fixed

File: kong/tracing/propagation.lua (hunk 5 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: fixed

File: kong/tracing/propagation.lua (hunk 6 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: fixed

**Comment by jschmid1** [comment]
@pathak1515 I'd like to re-emphasize the need for a unit-test. The reference issue https://github.com/Kong/kong/issues/9156 identified edge-cases which could be used as a starting point 👍

File: kong/tracing/propagation.lua [hunk 7]
```diff
@@ -41,7 +42,7 @@ local function parse_baggage_headers(headers, header_pattern)
   -- account for both ot and uber baggage headers
   local baggage
   for k, v in pairs(headers) do
-    local baggage_key = match(k, header_pattern)
+    local baggage_key = find(k, header_pattern, 1, true)
```
Comment (flrgh) [inline kong/tracing/propagation.lua:45]: This doesn't seem quite right. Using `string.find` this way performs a plain substring search, but the [possible inputs](https://github.com/Kong/kong/blob/8910980b2dcc580c6cd93130799b06115a6b853f/kong/tracing/propagation.lua#L21-L22) are all patterns and not plain strings. Also, `string.find` returns the position of the match, but the surrounding context makes it clear that we are interested in the matched substring itself (hence `string.match` being used).

**Review by flrgh — DISMISSED** [review]
Seems that the proper fix would be to correctly escape the offending pattern (`^ot-baggage%-(.*)$` => `^ot%-baggage%-(.*)$`) instead of using `string.find`.

File: kong/tracing/propagation.lua (hunk 7 above)
Comment (pathak1515) [reply kong/tracing/propagation.lua]: fixed

**Comment by pathak1515** [comment]
@mayocream @flrgh Can you folks please review the PR now?

**Comment by pathak1515** [comment]
@mayocream @flrgh Can you review the PR please?

**Comment by mayocream** [comment]
> @mayocream @flrgh Can you review the PR please?

Hi, this PR seem to be missing a unit test to confirm the fix.

**Comment by pathak1515** [comment]
@mayocream @flrgh Added unit tests to validate changes

**Comment by mayocream** [comment]
LGTM, thanks for adding the test.
Would you like to edit the Changelog.md file to add an entry describing this PR under "Fixes" section?

**Comment by pathak1515** [comment]
@mayocream added changelog.md entry for the same

**Review by mayocream — APPROVED** [review]

**Comment by pathak1515** [comment]
@flrgh Can you approve the PR here?

**Comment by pathak1515** [comment]
@mayocream Can we please get this PR merged?

**Comment by mayocream** [comment]
> @mayocream Can we please get this PR merged?

I'll ping someone with the merge permission. Don't worry, this should included in the next patch release for 3.0.

Thanks for your contribution（╹◡╹）

**Comment by flrgh** [comment]
> @flrgh Can you approve the PR here?

:wave: @pathak1515 thank you very much for your contribution. There is no need to spam this thread. The maintainers will get to it in due time.

File: spec/01-unit/26-tracing/02-propagation_spec.lua [hunk 8]
```diff
@@ -559,6 +559,33 @@ describe("propagation.parse", function()
...
+    local t = { parse({
+      ["ot-tracer-traceid"] = trace_id,
+      ["ot-tracer-spanid"] = span_id,
+      ["ot-tracer-sampled"] = "1",
+      ["ottttttttbaggage-foo"] = "invalid header"
+    })}
+    local mock_baggage_index = t[6]
+    assert.same({ "ot", trace_id, nil, span_id, true }, to_hex_ids(t))
+    assert.same(mock_baggage_index, nil)
+    assert.spy(warn).not_called()
```
Comment (mayocream) [inline spec/01-unit/26-tracing/02-propagation_spec.lua:588]:
```suggestion
    it("valid trace_id, valid span_id, sampled, invalid baggage added", function()
      local t = { parse({
        ["ot-tracer-traceid"] = trace_id,
        ["ot-tracer-spanid"] = span_id,
        ["ot-tracer-sampled"] = "1",
        ["ottttttttbaggage-foo"] = "invalid header"
      })}
      local mock_baggage_index = t[6]
      assert.same({ "ot", trace_id, nil, span_id, true }, to_hex_ids(t))
      assert.same(mock_baggage_index, nil)
      assert.spy(warn).not_called()
    end)
```

**Comment by kikito** [comment]
@mayocream please cherry-pick this into `release/3.0.1`

**Review by fffonion — APPROVED** [review]

**Review by dndx — APPROVED** [review]

**Comment by dndx** [comment]
Merged, thanks for your contribution @pathak1515 !
